### PR TITLE
Do not expect PDB files for a refonly build

### LINKED ISSUE #6510: Do not expect PDB files to be produced by a refonly build
### Issue Description

When the `ProduceOnlyReferenceAssembly` property is set to `true`, the `/refonly` switch is passed to the C# compiler, which then produces a reference assembly _instead of_ the normal build output. This never produces a .pdb file, regardless of the `DebugType` property.

However, MSBuild will currently still set `_DebugSymbolsProduced` to `true` unless `DebugType` happens to be `none` or `embedded`.

This then causes nuget packing to fail, because nonexistent .pdb files are reported by `DebugSymbolsProjectOutputGroup` and passed to the pack task. See also dotnet/sdk#17993.

### Steps to Reproduce

1. `dotnet new classlib`
2. Add to the project:
   ```xml
     <PropertyGroup>
       <!-- This defaults to 'true' with .NET 5+, which conflicts with ProduceOnlyReferenceAssembly. -->
       <ProduceReferenceAssembly>false</ProduceReferenceAssembly>
     </PropertyGroup>

     <Target Name="Foo" AfterTargets="Build" DependsOnTargets="DebugSymbolsProjectOutputGroup">
       <Error Text="Build Is Incorrectly Expected To Have Produced: %(DebugSymbolsProjectOutputGroupOutput.FullPath)"
              Condition=" !Exists('%(DebugSymbolsProjectOutputGroupOutput.FullPath)') " />
     </Target>
   ```
3. `dotnet build -p:ProduceOnlyReferenceAssembly=true`

### Expected Behavior

The `Foo` target raises no error.

### Actual Behavior

The `Foo` target reports an error.

### PR DESCRIPTION
**Pull request by Zastai**
Fixes #6510

### Context

MSBuild currently incorrectly expects PDB files to be produced by a refonly compile.

### Changes Made

This defaults both `ProduceReferenceAssembly` and `ProduceOnlyReferenceAssembly` (slightly earlier than before).

When `ProduceOnlyReferenceAssembly` is set to `true`, don't expect symbols to be produced.

### CODE REVIEW

**Review by rainersigwald — COMMENTED** [review]
Seems reasonable to me but can I get you to nominate a compiler reviewer @jaredpar?

**Comment by Zastai** [comment]
Just realised I created this against main, but it would probably make sense to retarget it to vs16.11. It's probably not critical enough to warrant a vs16.10 fix.

**Comment by rainersigwald** [comment]
Yeah, this wouldn't meet the 16.10 bar, but 16.11 makes sense to me too--and if we decide otherwise it's easy to merge from there to main, so please rebase.

**Comment by jaredpar** [comment]
@chsienki, @jcouv PTAL

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 1]
```diff
@@ -169,6 +169,10 @@ Copyright (C) Microsoft Corporation. All rights reserved.
     <OutputType Condition=" '$(TargetType)' == 'Container' or '$(TargetType)' == 'DocumentContainer' ">library</OutputType>
     <OutputType Condition=" '$(OutputType)' == '' ">exe</OutputType>
 
+    <!-- Whether or not a reference assembly (or ONLY a reference assembly) is produced. -->
+    <ProduceReferenceAssembly Condition="'$(ProduceReferenceAssembly)' == ''">false</ProduceReferenceAssembly>
+    <ProduceOnlyReferenceAssembly Condition="'$(ProduceOnlyReferenceAssembly)' == ''">false</ProduceOnlyReferenceAssembly>
```
Comment (jcouv) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:174]: Is this change necessary?

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 1). Just one minor question

File: src/Tasks/Microsoft.Common.CurrentVersion.targets (hunk 1 above)
Comment (Zastai) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:174]: Well, the next change tests `ProduceOnlyReferenceAssembly` so defaulting it seemed reasonable. And it seemed equally reasonable to then have the defaulting of `ProduceReferenceAssembly` in the same place.

**Review by jcouv — APPROVED** [review]
LGTM Thanks

**Review by Forgind — APPROVED** [review]

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 2]
```diff
@@ -179,13 +183,11 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
     <_DebugSymbolsProduced Condition="'$(DebugType)'=='portable'">true</_DebugSymbolsProduced>
     <_DebugSymbolsProduced Condition="'$(DebugType)'=='embedded'">false</_DebugSymbolsProduced>
+    <_DebugSymbolsProduced Condition="'$(ProduceOnlyReferenceAssembly)'=='true'">false</_DebugSymbolsProduced>
 
     <!-- Whether or not a .xml file is produced. -->
     <_DocumentationFileProduced>true</_DocumentationFileProduced>
     <_DocumentationFileProduced Condition="'$(DocumentationFile)'==''">false</_DocumentationFileProduced>
-
-    <!-- Whether or not a reference assembly is produced. -->
-    <ProduceReferenceAssembly Condition="'$(ProduceReferenceAssembly)' == ''">false</ProduceReferenceAssembly>
```
Comment (Ely68) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:188]: Whether or not a reference assembly is produced

**Review by Ely68 — CHANGES_REQUESTED** [review]

**Comment by Ely68** [comment]
all checks pass

**Review by Ely68 — APPROVED** [review]

File: src/Tasks/Microsoft.Common.CurrentVersion.targets (hunk 1 above)
Comment (rainersigwald) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:174]: This isn't _necessary_ but is also generally fine.

In MSBuild, there are no true booleans, only strings treated like booleans. As a result, each property that is logically a boolean can have three states:

1. `true`
2. `false`
3. Empty, unset, or set to something else.

if you control the code that reads a variable, you can use this to your advantage by acting as though it's got a "default":
|condition|effect
|----------------------------|-----------------------|
| `'$(Property)' == 'false'` | "Defaults" to `true`  |
| `'$(Property)' == 'true'`  | "Defaults" to `false` |

Since this isn't obvious, it's often a good idea to explicitly declare a default value, as is done here.

There's a downside, though, which is that explicitly declaring a value makes it harder for the user to have a default value of their own, defined with `Condition="'$(ProduceOnlyReferenceAssembly)' == ''"`. It's possible that users might have used that condition in `Directory.Build.targets` or NuGet `.targets`.

I don't see any of those in a [search for `ProduceOnlyReferenceAssembly extension:targets`](https://github.com/search?l=&p=2&q=ProduceOnlyReferenceAssembly+extension%3Atargets&ref=advsearch&type=Code), but there are too many hits to be totally confident.

As a result, and since this is targeting a release that is intended to minimize risk, **I think we should back out these changes and add only the `_DebugSymbolsProduced` check.**

**Review by rainersigwald — CHANGES_REQUESTED** [review]

File: src/Tasks/Microsoft.Common.CurrentVersion.targets (hunk 1 above)
Comment (Zastai) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:174]: That's fair enough.

For boolean properties, is there a Microsoft convention and/or best practice for testing them? Always comparing to `'true'`? Always comparing to `'false'`? Always using equal tests for `'true'/'false'`? (The latter is probably something to avoid, because it would potentially match neither condition.)

And yes, it would be convenient to have some means of safely specifying property defaults in a .props file without also assigning at that point. Setting a bunch of properties in .targets always feels a bit naughty :)

Comment (Zastai) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets]: Done

**Review by rainersigwald — APPROVED** [review]

**Comment by rainersigwald** [comment]
Thanks, @Zastai!
